---
title: "Nucleosome Positioning"
output:
  BiocStyle::html_document:
    toc: true
bibliography: rjmcmc.bibtex
vignette: >
  %\VignetteIndexEntry{Nucleosome Positioning}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignettePackage{RJMCMC}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
library(knitr)
```

<br />
**Package**: `r Rpackage("RJMCMC")`<br />
**Authors**: `r packageDescription("RJMCMC")[["Author"]]`<br />
**Version**: `r packageDescription("RJMCMC")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("RJMCMC")[["License"]]`<br />

# Licensing and citing

The **RJMCMC** package and the underlying **RJMCMC** code are distributed 
under the Artistic license 2.0. You are free to use and redistribute 
this software. 

If you use this package for a publication, we would ask you to cite the
following:

>Samb R, Khadraoui K, Belleau P, et al. (2015) Using informative Multinomial-Dirichlet prior in a t-mixture with reversible jump estimation of nucleosome positions for genome-wide profiling. Statistical Applications in Genetics and Molecular Biology. Volume 14, Issue 6, Pages 517-532, ISSN (Online) 1544-6115, ISSN (Print) 2194-6302, December 2015, <a href="http://dx.doi.org/10.1515/sagmb-2014-0098">doi:10.1515/sagmb-2014-0098</a>


# Introduction

Global gene expression patterns are established and maintained by the 
concerted actions of transcription factors (TFs) and the proteins that 
constitute chromatin. The key structural element of chromatin is the 
nucleosome, which consists of an octameric histone core wrapped by 146 bps 
of DNA and connected to its neighbor by approximately 10-80 pbs of linker 
DNA [@Polishko2012].

The literature on nucleosome positioning commonly focuses on frequentist 
inferences within parametric approaches (see for instance @Chen2010 and
@Xi2010).  In those works, the detection of nucleosome positions is done 
using a hidden Markov model with a known order by 
assumption.  

The **RJMCMC** package is an implemantation of a fully Bayesian hierarchical 
model for profiling of nucleosome positions based on high-throughput
short-read data (MNase-Seq data). The implemantation is based on a strategy 
which incorporates four aspects. First, it jointly models local concentrations
of directional reads. Second, it uses a Multinomial-Dirichlet model in the 
construction of an informative prior distribution coupled to a t-mixture model 
with unknown degrees of freedom. Third, the number of nucleosomes
is considered to be a random variable and refers to a prior distribution. 
Fourth, the unknown parameters are simultaneously using the reversible jump 
Markov chain Monte Carlo (RJMCMC) simulation technique (see for instance 
@Green1995 and @Richardson1997). 

Detailed information about the **RJMCMC** model can be found in the article 
mentioned in the precedent section.

# Loading the RJMCMC package

As with any R package, the **RJMCMC** package should first be loaded with 
the following command:

```{r loadingPackage, warning=FALSE, message=FALSE} 
library(RJMCMC)
```


# RJMCMC analysis

A typical **RJMCMC** analysis consists of the following steps:

1. Segment the genome into candidate regions that have sufficient aligned
reads. The region cannot be larger than one chromosome.
2. Estimate nucleosome positions for each region.
3. Regroup all regions together. The final region cannot be larger than
one chromosome.
4. Post-process predictions of the regrouped region to correct certain 
predictions.

# RJMCMC analysis step by step

A synthetic nucleosome sample containing 300 nucleosomes has been created 
for the purposes of this example.The Bioconductor package 
`r Biocpkg("nucleoSim")`.

```{r }
## Load nucleoSim package
library(nucleoSim)

val.num       <- 300  ### Number of well-positioned nucleosomes
val.del       <- 50   ### Number of well-positioned nucleosomes to delete
val.var       <- 30   ### variance associated to the well-positioned nucleosomes
val.fuz       <- 50   ### Number of fuzzy nucleosomes
val.fuz.var   <- 50   ### variance associated to the fuzzy nucleosomes
val.rnd.seed  <- 444  ### Set seed when result needs to be reproducible
val.offset    <- 10000  ### The number of bases used to offset all nucleosomes and reads

set.seed(292)   ### Set seed when result needs to be reproducible

## Create sample using a Normal distribution
sample <- nucleoSim::syntheticNucReadsFromDist(wp.num=val.num,
                                    wp.del=val.del,
                                    wp.var=val.var,
                                    fuz.num=val.del,
                                    fuz.var=val.fuz.var,
                                    rnd.seed=val.rnd.seed,
                                    distr="Normal",
                                    offset=val.offset)

## Create visual representation of the synthetic nucleosome sample
plot(sample)
```

## Segment the analyzed region into candidate regions

A genome can be pre-processed and segmented into candidate regions to 
accelerate the analysis. 

A genome must at least be separated by chromosome since the RJMCMC 
algorithm can only analyze one chromosome at the time.

Genome segmentation can be done using the `segmentPING()` function 
from the `r Biocpkg("PING")` package.

```{r segment01, warning=FALSE, collapse=TRUE, message=FALSE} 
## Load PING package
library("PING")

## Transform sample dataset into GRanges object
sampleGRanges <- GRanges(seqnames = sample$dataIP$chr, 
                          ranges = IRanges(start = sample$dataIP$start, 
                                           end = sample$dataIP$end), 
                          strand = sample$dataIP$strand)

## Segment sample
sampleSegmented <- PING::segmentPING(sampleGRanges)

## Number of segments created
length(sampleSegmented@List)
```

## Run RJMCMC on each segment (when segmented)

```{r demo01 , warning=FALSE, collapse=TRUE} 
## Run first segment
readsForward <- sampleSegmented@List[[1]]@yF
readsReverse <- sampleSegmented@List[[1]]@yR

## Run RJMCMC analysis
## A higher number of iterations is recommanded for real analysis
resultSegment01 <- rjmcmc(startPosForwardReads = readsForward,
            startPosReverseReads = readsReverse,
            nbrIterations = 1000, lambda = 3, kMax = 30,
            minInterval = 100, maxInterval = 200, minReads = 5)

# Print the resulting nucleosomes and the related information
resultSegment01

```


# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```


# References
