---
title: "RJMCMC: Nucleosome Positioning"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Detection of consensus peak regions inside a group of experiences using narrowPeak files}
  %\VignettePackage{consensusSeekeR}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

</br>

Author: Rawane Samb, Astrid Louise Deschenes, Khader Khadraoui, Lajmi Lakhal-Chaieb, 
Pascal Belleau and Arnaud Droit<br />
Compilation date: `r Sys.Date()`

## 1. Licensing and citing

This package and the underlying RJMCMC code are distributed 
under the Artistic license 2.0. You are free to use and redistribute 
this software. 

If you use this package for a publication, we would ask you to cite the
following:

> Samb R, Desch\^enes A, Khadraoui K, Lakhal L, Belleau P et Droit A. Using 
> informative Multinomial-Dirichlet prior in a t-mixture with reversible jump 
> estimation of nucleosome positions for genome-wide profiling. 
> Submitted (2015). 

## 2. Introduction

Global gene expression patterns are established and maintained by the 
concerted actions of Transcription Factors (TFs) and the proteins that 
constitute chromatin. The key structural element of chromatin is the 
nucleosome, which consists of an octameric histone core wrapped by 146 bps 
of DNA and connected to its neighbor by approximately 10-80 pbs of linker 
DNA (reference TODO).

## 3. Loading the RJMCMC package

As with any R package, the **RJMCMC** package should first be loaded with 
the following command:

```{r loadingPackage, warning=FALSE, message=FALSE} 
library(rjmcmc)
```

## 4. RJMCMC analysis

```{r demo01 , collapse=TRUE} 
## Load the dataset containing forward and reverse reads
data(reads_demo)

## Nucleosome positioning
result <- RJMCMC(startPosForwardReads = reads_demo$readsForward,
            startPosReverseReads = reads_demo$readsReverse,
            nbrIterations = 1000, lambda = 3, kMax = 30,
            minInterval = 140, maxInterval = 295, minReads = 5)

# Print the resulting nucleosomes and the related information
result
```

## 5. References

TODO
